File: crates/neo4r-client/examples/basic_usage.rs
use neo4r_client::{Client, ClientError, QueryParams, QueryRow, QueryValue, Value};

fn main() -> Result<(), ClientError> {
    let address = std::env::args()
        .nth(1)
        .unwrap_or_else(|| "127.0.0.1:17687".to_string());
    let mut client = match Client::connect(&address) {
        Ok(client) => client,
        Err(err) => {
            eprintln!("failed to connect to neo4r at {address}: {err}");
            eprintln!(
                "start a local server first:\n  cargo run -p neo4r-server -- --bind {address} --data-dir /tmp/neo4r-rust-sdk-example --shards 1 --partitions 1"
            );
            return Err(err);
        }
    };

    client.ping()?;

    let mut params = QueryParams::new();
    params.insert("name".to_string(), Value::String("Alice".to_string()));
    params.insert("age".to_string(), Value::Int(42));
    let rows = client.execute_with_params(
        "CREATE (n:Person {name: $name, age: $age}) RETURN n.name, n.age",
        &params,
    )?;
    println!(
        "created: {} {}",
        scalar_string(&rows[0], "n.name")?,
        scalar_int(&rows[0], "n.age")?
    );

    let mut params = QueryParams::new();
    params.insert("name".to_string(), Value::String("Alice".to_string()));
    let rows = client.query_with_params(
        "MATCH (n:Person) WHERE n.name = $name RETURN n.name, n.age",
        &params,
    )?;
    println!(
        "matched: {} {}",
        scalar_string(&rows[0], "n.name")?,
        scalar_int(&rows[0], "n.age")?
    );

    let profile = client.profile("MATCH (n:Person) RETURN n", &QueryParams::new())?;
    println!("profile: {profile}");

    let plan = client.query_plan("MATCH (n:Person) RETURN n", &QueryParams::new())?;
    println!("query_plan: {plan}");

    let status = client.storage_status()?;
    println!("storage_status: {status}");

    let cluster = client.cluster_status()?;
    println!("cluster_status: {cluster}");

    client.close()
}

fn scalar_string(row: &QueryRow, key: &str) -> Result<String, ClientError> {
    match row.get(key) {
        Some(QueryValue::Scalar(Value::String(value))) => Ok(value.clone()),
        value => Err(ClientError::Protocol(format!(
            "{key} is not a string scalar: {value:?}"
        ))),
    }
}

fn scalar_int(row: &QueryRow, key: &str) -> Result<i64, ClientError> {
    match row.get(key) {
        Some(QueryValue::Scalar(Value::Int(value))) => Ok(*value),
        value => Err(ClientError::Protocol(format!(
            "{key} is not an int scalar: {value:?}"
        ))),
    }
}

File: sdks/python/examples/basic_usage.py
from __future__ import annotations

import argparse
import socket
import sys

from neo4r_client import Client, QueryValue


def scalar(row: dict[str, QueryValue], key: str):
    value = row[key]
    if value.kind != "scalar":
        raise TypeError(f"{key} is not a scalar value: {value!r}")
    return value.value


def main() -> None:
    parser = argparse.ArgumentParser(description="Run a basic neo4r Python SDK example.")
    parser.add_argument("--host", default="127.0.0.1")
    parser.add_argument("--port", default=17687, type=int)
    args = parser.parse_args()

    try:
        client = Client.connect(args.host, args.port)
    except OSError as err:
        print(
            f"failed to connect to neo4r at {args.host}:{args.port}: {err}",
            file=sys.stderr,
        )
        print(
            "start a local server first:\n"
            "  cargo run -p neo4r-server -- "
            f"--bind {args.host}:{args.port} "
            "--data-dir /tmp/neo4r-python-sdk-example "
            "--shards 1 --partitions 1",
            file=sys.stderr,
        )
        raise SystemExit(1)

    try:
        client.ping()
        rows = client.execute(
            "MERGE (n:Person {sample_id: $sample_id}) "
            "ON CREATE SET n.name = $name, n.age = $age "
            "ON MATCH SET n.name = $name, n.age = $age "
            "RETURN n.name, n.age",
            {"sample_id": "basic-usage", "name": "Alice", "age": 42},
        )
        print("created:", scalar(rows[0], "n.name"), scalar(rows[0], "n.age"))

        rows = client.query(
            "MATCH (n:Person) WHERE n.sample_id = $sample_id RETURN n.name, n.age",
            {"sample_id": "basic-usage"},
        )
        print("matched:", scalar(rows[0], "n.name"), scalar(rows[0], "n.age"))

        profile = client.profile("MATCH (n:Person) RETURN n", {})
        print("profile:", profile)

        plan = client.query_plan("MATCH (n:Person) RETURN n", {})
        print("query_plan:", plan)

        status = client.storage_status()
        print("storage_status:", status)

        cluster = client.cluster_status()
        print("cluster_status:", cluster)
    finally:
        client.close()


if __name__ == "__main__":
    main()
